graphstorm-processing/graphstorm_processing/repartition_files.py

```diff
@@ -0,0 +1,997 @@
+    )
+
+    return parser.parse_args(args)
+
+
+def main():
+    """
+    Re-partitions the output of the distributed processing pipeline
+    to ensure all files that belong to the same edge/node type have the same
+    number of rows per corresponding part-file.
+
+    This is done by reading the metadata file and partitioning the files
+    according to the row counts of the features in the metadata file.
+
+    The output is written to the same location as the input, with the
+    updated metadata file.
+
+    Parameters
+    ----------
+    input_prefix : str
+        Prefix path to where the output was generated
+        from the distributed processing pipeline.
+        Can be a local path or S3 prefix (starting with 's3://').
+    metadata_file_name : str
+        Name of the original partitioning pipeline metadata file.
+    updated_metadata_file_name : str
+        Name of the updated (output) partitioning pipeline metadata file.
+    log_level : str
+        Log level.
+
+    Notes
+    -----
+    This function is meant to be used as a command-line tool.
+    """
+    args = parse_args(sys.argv[1:])
+    logging.basicConfig(level=getattr(logging, args.log_level.upper(), None))
+    if args.input_prefix.startswith("s3://"):
+        filesystem_type = "s3"
+    else:
+        input_prefix = str(Path(args.input_prefix).resolve(strict=True))
+        filesystem_type = "local"
+
+    # Trim trailing '/' from S3 URI
+    if filesystem_type == "s3":
+        input_prefix = s3_utils.s3_path_remove_trailing(args.input_prefix)
+
+    logging.info(
+        "Re-partitioning files under %s to ensure all files that belong to the same "
+        "edge/node type have the same number of rows per part-file.",
+        input_prefix,
+    )
+
+    t0 = time.time()
+
+    logging.info("Reading metadata from %s/%s", input_prefix, args.metadata_file_name)
+
+    # Get the metadata file
+    region = None
+    if filesystem_type == "s3":
+        bucket = input_prefix.split("/")[2]
+        s3_key_prefix = input_prefix.split("/", 3)[3]
+        region = s3_utils.get_bucket_region(bucket)
+        s3_client = boto3.client("s3", region_name=region)
+        s3_client.download_file(
+            bucket, f"{s3_key_prefix}/{args.metadata_file_name}", f"/tmp/{args.metadata_file_name}"
+        )
+        metadata_filepath = os.path.join("/tmp/", args.metadata_file_name)
+    else:
+        metadata_filepath = os.path.join(input_prefix, args.metadata_file_name)
+
+    metadata_filepath = str(Path(metadata_filepath).resolve(strict=True))
+
+    with open(metadata_filepath, "r", encoding="utf-8") as metafile:
+        metadata_dict = json.load(metafile)  # type: Dict[str, Dict]
+
+    edge_structure_meta = metadata_dict["edges"]  # type: Dict[str, Dict[str, Dict]]
+
+    task_type = metadata_dict["graph_info"]["task_type"]  # type: str
+
+    edge_data_exist = "edge_data" in metadata_dict.keys() and metadata_dict["edge_data"]
```

Comment:
Does this mean: 1) "edge_data" in metadata_dict.keys() and metadata_dict["edge_data"] is not None? Shall we check len(metadata_dict["edge_data"]) > 0?

Reply:
I mean we add another check after checking ```metadata_dict["edge_data"] is not None```.

Reply:
I think the bool check `metadata_dict["edge_data"]` covers the case where the dict is empty and/or missing. See:  https://docs.python.org/3/library/stdtypes.html#truth-value-testing

```
>>> d = {}
>>> len(d) > 0
False
>>> bool(d)
False
>>> d = None
>>> bool(d)
False
```